File: baekjoononlinejudge/1654/1654.py
import sys
input = sys.stdin.readline

K, N = map(int, input().split())
lans = [int(input()) for _ in range(K)]

left = 1
right = max(lans)

while left <= right:
    processdLans = 0
    mid = (right + left) // 2

    for lan in lans:
        processdLans += lan // mid

    if N <= processdLans:
        result = mid
        left = mid + 1
    else:
        right = mid - 1
        
print(result)

File: BaekjoonOnlineJudge/1654/1654.java
import java.io.*;
import java.util.*;


public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static int k, n;
    private static int[] lanCables;

    private static long getLanCable(long length) {
        return Arrays.stream(lanCables).mapToLong(lanCable -> lanCable / length).sum();
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        k = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        lanCables = new int[k];

        for (int i = 0; i < k; i++) {
            lanCables[i] = Integer.parseInt(br.readLine());
        }

        long answer = 0;
        long left = 1;
        long right = Arrays.stream(lanCables).max().getAsInt();

        while (left <= right) {
            long mid = (left + right) / 2;

            int cutLanCable = (int) getLanCable(mid);

            if (cutLanCable < n) {
                right = mid - 1;
            } else {
                answer = Math.max(answer, mid);
                left = mid + 1;
            }
        }

        bw.write(String.valueOf(answer));

        bw.flush();
        bw.close();
        br.close();
    }
}